blog/modules/ROOT/pages/7-graal-native-executables.adoc

```diff
@@ -0,0 +1,475 @@
+
+_Executable Assembly_
+```bash
+$ ./out/foo/assembly.dest/out.jar --text hello-world
+135
+170
+178
+188
+191
+192
+192
+189
+190
+188
+195
+185
+182
+```
+
+_Native Image_
+```bash
+$ time ./out/foo/nativeImage.dest/native-executable --text hello-world
+171
+163
+180
+173
+182
+182
+181
+184
+181
+181
+182
+183
+181
+```
+
+As you can see, the executable assembly and native image both have comparable performance,
+although the executable assembly starts off lower (135 vs 171) for the first iteration due
+to JVM warmup time, but eventually reaches a higher steady state than the native image
+(~190 vs ~180).
+
+While again this is for a toy program, for larger applications the same pattern applies: Graal
+native binaries avoid the slow startup that JVM applications often exhibit, but in exchange
+may not quite reach the same peak steady-state performance that a long-lived JVM application
+would typically achieve.
+
+### Memory Usage
+
+While our programs are looping, we can also see how much memory they take via `top`:
+
+_Executable Assembly_
+```bash
+$ jps
+58547 MillMain
+86276 MillServerMain
+24895 Jps
+9263 Foo
+1071 Main
+
+$ top | grep 9263
+9263   java             0.0  00:20.41 32/1   1   134    373M  0B    0B    9263  42892 running  *0[1]       0.00000 0.00000    501 93089     9569   5005      2470      387381     104652     75938      9       0        0.0   0      0      lihaoyi            N/A    N/A   N/A   N/A   N/A   N/A
+```
+
+_Native Image_
+```bash
+$ ps aux | grep native-executable
+lihaoyi          43880  46.1  0.1 408681792  30176 s000  S+    3:40PM   0:05.84 ./out/foo/nativeImage.dest/native-executable --text hello-world
+lihaoyi          86276   0.0  2.1 420349904 720416 s000  S     3:14PM   1:00.88 /Library/Java/JavaVirtualMachines/amazon-corretto-17.jdk/Contents/Home/bin/java -cp /Users/lihaoyi/.cache/mill/download/0.12.5-68-e4bf78 mill.runner.MillServerMain /Users/lihaoyi/Github/mill/blog/modules/ROOT/attachments/7-graal-native-executables/out/mill-server/aa508f0984fd2811f6c6d8fae1362f1774e4f5f7-1
+lihaoyi          48496   0.0  0.0 408626896   1376 s002  S+    3:40PM   0:00.00 grep native-executable
+
+$ top | grep 43880
+43880  native-executabl 0.0  00:10.19 3/1    0   26     20M   0B    0B    43880 42892 running  *0[1]       0.00000 0.00000    501 695907    44380  8100      4045      153233     8177       24637      313     0        0.0   0      0      lihaoyi            N/A    N/A   N/A   N/A   N/A   N/A
+```
+
+The column `373M` and `20M` are the respective memory footprints of the executable assembly
+and native image binary. In this small program, the native image uses almost 20x less memory
+than the JVM executable assembly! That is a very significant reduction in resource footprint
+
+### Portability and Hermeticity
+
+Executable assembly jars require a JVM installed globally in order to run. In a way they are
+not hermetic, since the globally-installed JVM can differ resulting in the assembly behaving
```

Comment:
Sentence is odd/hard to read with two words for 'difference' and passive voice.

My slight improvement, making it two sentences and reducing passive voice, so that it has less going on

`... not hermitic. A different globally-installed JVM can result in the assembly behaving different at runtime`.